### PR TITLE
Migration Angular 11

### PR DESCRIPTION
**Pull request by LucasBrazi06**

### CODE REVIEW

**Comment by LucasBrazi06** [comment]
@jerome-benoit 
Let me know your feedback.

Migration from Angular 9 to 11
Refactored module dependencies
Fixed errors appearing in console
Cut cross module dependencies in table action: moved them back in shared folder
Use of relative path in import of modules

-> Now the UI is much more responsive in dev, Search field is working again !!!
-> We gained also size at first load
-> We have now the size of each modules after compilation so we can enhance loading time

Still actions to do like to remove the usage of form 'disabled' in templates (issue opened already)

I'll merge tomorrow after few other tests.
Let me know.

**Comment by LucasBrazi06** [comment]
Prod:

**Comment by jerome-benoit** [comment]
> @jerome-benoit
> Let me know your feedback.
> 
> Migration from Angular 9 to 11
> Refactored module dependencies
> Fixed errors appearing in console
> Cut cross module dependencies in table action: moved them back in shared folder
> Use of relative path in import of modules
> 
> -> Now the UI is much more responsive in dev, Search field is working again !!!
> -> We gained also size at first load
> -> We have now the size of each modules after compilation so we can enhance loading time
> 
> Still actions to do like to remove the usage of form 'disabled' in templates (issue opened already)
> 
> I'll merge tomorrow after few other tests.
> Let me know.

Just tell me how you've made the update: 
+ commands used; 
+ documentation used;

File: .gitignore [hunk 1]
```diff
@@ -28,27 +35,17 @@ manifest.yml
 !.vscode/tasks.json
 !.vscode/launch.json
 !.vscode/extensions.json
-
-# Vim
-# Persistent undo
-[._]*.un~
-# Temporary
-.netrwhist
-*~
```
Comment (jerome-benoit) [inline .gitignore:37]: Please keep the VIM related files to ignore.

File: .gitignore [hunk 2]
```diff
@@ -28,27 +35,17 @@ manifest.yml
...
 /coverage
 /libpeerconnection.log
 npm-debug.log
-npm-debug.log*
 yarn-error.log
-/yarn.lock
 testem.log
 /typings
-/output
-*.orig
```
Comment (jerome-benoit) [inline .gitignore:51]: Keep *.orig to ignore merge artefacts.

File: package.json [hunk 3]
```diff
@@ -35,7 +35,8 @@
     "git:sreset": "git submodule foreach 'git reset --hard'",
     "ngcc": "ngcc",
     "ng": "ng",
-    "start": "npm run ngcc && ng serve --configuration=development",
+    "start": "ng serve --configuration=development",
+    "postinstall": "npm run ngcc",
```
Comment (jerome-benoit) [inline package.json:39]: Double check that it will not break xmake.

**Review by jerome-benoit — CHANGES_REQUESTED** [review]
I think you do not have done material update associated to the angular update. 
The upgrade path is in three parts:
+ npm run angular-update -> check and commit
+ npm run material-update -> check and commit
+ npm run angular-dependencies-update -> check and commit 

You can also update the script for angular same major version related update 9->11: 

 "angular9-update": "ng update @angular/core@9 @angular/cli@9",
 "material9-update": "ng update @angular/material@9",

And the import sorting seems incorrect, make sure it's done properly by running:
+ npm run import-sort
+ npm run tslint:fix
to avoid to generate conflicts for nothing relevant. 

That PR should also be merged after all other PRs; otherwise all devs will have to resolve huge conflicts during the next week. Please avoid to give them that extra work.

File: package.json [hunk 4]
```diff
@@ -59,72 +60,66 @@
...
     "ngx-captcha": "^8.0.1",
-    "ngx-daterangepicker-material": "^3.0.4",
+    "ngx-daterangepicker-material": "^4.0.1",
     "ngx-google-places-autocomplete": "^2.0.5",
-    "ngx-moment": "^3.5.0",
-    "perfect-scrollbar": "^1.5.0",
+    "ngx-moment": "^5.0.0",
     "popper.js": "^1.16.1",
-    "rxjs": "^6.6.3",
-    "socket.io-client": "^2.3.1",
```
Comment (jerome-benoit) [inline package.json:106]: "socket.io-client": "^2.3.1",
Do not remove that dep ;)

File: package.json [hunk 5]
```diff
@@ -150,9 +146,9 @@
     "puppeteer": "^3.3.0",
     "rimraf": "^3.0.2",
     "rosie": "^2.0.1",
-    "ts-node": "^8.10.2",
-    "tslint": "^6.1.3",
-    "typescript": "~3.8.3",
+    "ts-node": "~8.3.0",
+    "tslint": "~6.1.0",
```
Comment (jerome-benoit) [inline package.json:150]: Do not downgrade tslint.

File: src/app/app.component.spec.ts [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    }).compileComponents();
+  });
+
+  it('should create the app', () => {
+    const fixture = TestBed.createComponent(AppComponent);
+    const app = fixture.componentInstance;
+    expect(app).toBeTruthy();
+  });
+
+  it(`should have as title 'ev-dashboard-new'`, () => {
```
Comment (jerome-benoit) [inline src/app/app.component.spec.ts:23]: The app name is ev-dashboard

**Comment by LucasBrazi06** [comment]
> Just tell me how you've made the update:
> 
> * commands used;
> * documentation used;

I simply created an Angular 11 project (ng new) and moved the src from the old dashboard to the new one.
Material has been installed with the ng command also.

File: .editorconfig [hunk 7]
```diff
@@ -1,15 +1,16 @@
...
-trim_trailing_whitespace = true
 insert_final_newline = true
-end_of_line = lf
-max_line_length = 100
+trim_trailing_whitespace = true
+
+[*.ts]
+quote_type = single
 
-[{Makefile,**.mk}]
```
[the remark this answers is not in the document]
Comment (LucasBrazi06) [reply .editorconfig]: Fixed

File: .editorconfig [hunk 8]
```diff
@@ -1,15 +1,16 @@
...
+charset = utf-8
 indent_style = space
 indent_size = 2
-tab_width = 2
-charset = utf-8
-trim_trailing_whitespace = true
 insert_final_newline = true
-end_of_line = lf
-max_line_length = 100
+trim_trailing_whitespace = true
```
[the remark this answers is not in the document]
Comment (LucasBrazi06) [reply .editorconfig]: Fixed

File: .gitignore (hunk 1 above)
Comment (LucasBrazi06) [reply .gitignore]: Fixed

File: .gitignore (hunk 2 above)
Comment (LucasBrazi06) [reply .gitignore]: Fixed

File: package.json (hunk 3 above)
Comment (LucasBrazi06) [reply package.json:39]: I think that we will end up with two different repos the way I see things today with our colleagues.
First we have to move to SAP public space and next the repo will be forked to the private one where we will not have submodules anymore.
Let's see.
For the time being, let's not keep it in full sync with SAP one.

File: package.json (hunk 4 above)
Comment (LucasBrazi06) [reply package.json:106]: As it's working without I believed this was not necessary, as well as for rxjs...

File: package.json (hunk 5 above)
Comment (LucasBrazi06) [reply package.json:150]: It was set by Angular 11 setup, I'll change it.

Comment (LucasBrazi06) [reply package.json]: Done

File: package.json (hunk 4 above)
Comment (LucasBrazi06) [reply package.json:106]: Done

File: src/app/app.component.spec.ts (hunk 6 above)
Comment (LucasBrazi06) [reply src/app/app.component.spec.ts]: Fixed

**Comment by jerome-benoit** [comment]
> > Just tell me how you've made the update:
> > 
> > * commands used;
> > * documentation used;
> 
> I simply created an Angular 11 project (ng new) and moved the src from the old dashboard to the new one.
> Material has been installed with the ng command also.

Have a look at the official guide then: https://update.angular.io/?v=9.1-11.0 (check 'I use Angular Material'). 
And run the commands I've listed in the review (it's the same as in the guide but adapted to run in a project). 
You probably have to add a script for the material 10 intermediate upgrade and then run again the material update to go to the 11.
You can run the upgrade commands several times, it's harmless. The order is important and you have to commit any changes made in between.

Otherwise, it looks good. The angular process is a bit trickier when you go up 2 majors release.

File: package.json (hunk 4 above)
Comment (jerome-benoit) [reply package.json:106]: It probably have added as a sub-dependency of a dependency. But you then do not control the version used and it's important for socket.io.

File: package.json (hunk 3 above)
Comment (jerome-benoit) [reply package.json:39]: Ok, but before it's done I think it's better do not break the current setup. 
And doing an automated synchro between a public repo with submodules to a private one that will incorporate the submodules in the source is going to be difficult and very hard to maintain. Maybe not possible: git cannot handle two different history.
We probably have to discuss why we need an internal mirror and if it's really needed. If yes, find the simplest solution to mirror it automatically.

**Comment by jerome-benoit** [comment]
> You can run the upgrade commands several times, it's harmless. The order is important and you have to commit any changes made in between.

I can also do it on that branch if you agree but I prefer to not be the one that knows how to make an angular major update. So it's better to document it and let other do it.

**Comment by LucasBrazi06** [comment]
> Have a look at the official guide then: https://update.angular.io/?v=9.1-11.0 (check 'I use Angular Material').
> And run the commands I've listed in the review (it's the same as in the guide but adapted to run in a project).
> You probably have to add a script for the material 10 intermediate upgrade and then run again the material update to go to the 11.
> You can run the upgrade commands several times, it's harmless. The order is important and you have to commit any changes made in between.
> 
> Otherwise, it looks good. The angular process is a bit trickier when you go up 2 majors release.

It's not an upgrade, I created a new project from scratch and redo everything.

So after creating the Angular 11 project:
- ng new ev-dashboard

I added the material components:
- ng add @angular/material

Then I didn't have to take car of a migration at all :)

File: package.json (hunk 3 above)
Comment (LucasBrazi06) [reply package.json:39]: Ok, I'll discuss this with you next week :)

**Review by jerome-benoit — APPROVED** [review]

**Pull request merged**